client/db/bolt/upgrades.go

```diff
@@ -0,0 +1,122 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package bolt
+
+import (
+	"fmt"
+
+	"decred.org/dcrdex/dex/encode"
+	"go.etcd.io/bbolt"
+)
+
+const (
+	initialVersion = 0
+
+	// versionedDBVersion is the second version of the database. It versions
+	// the database by persisting the version.
+	versionedDBVersion = 1
+
+	// DBVersion is the latest version of the database that is understood by the
+	// program. Databases with recorded versions higher than this will fail to
+	// open (meaning any upgrades prevent reverting to older software).
+	DBVersion = versionedDBVersion
+)
+
+// upgrade the database to the next version. Each database upgrade function
+// should be keyed by the database version it upgrades.
+
+var upgrades = [...]func(tx *bbolt.Tx) error{
+	versionedDBVersion - 1: versionedDBUpgrade,
+}
+
+func fetchDBVersion(tx *bbolt.Tx) (uint32, error) {
+	bucket := tx.Bucket(appBucket)
+	if bucket == nil {
+		return 0, fmt.Errorf("app bucket not found")
+	}
+
+	versionB := bucket.Get(versionKey)
+	if versionB == nil {
+		return 0, fmt.Errorf("database version not found")
+	}
+
+	return encode.BytesToUint32(versionB), nil
+}
+
+func setDBVersion(tx *bbolt.Tx, newVersion uint32) error {
+	bucket := tx.Bucket(appBucket)
+	if bucket == nil {
+		return fmt.Errorf("app bucket not found")
+	}
+
+	return bucket.Put(versionKey, encode.Uint32Bytes(newVersion))
+}
+
+// upgradeDB checks whether any upgrades are necessary before the database is
+// ready for application usage.  If any are, they are performed.
+func upgradeDB(db *bbolt.DB) error {
```

Comment:
If I'm reading everything right, then every time a new DB is created with `NewDB`, it will be run through every upgrader, because it will have a `nil` entry at the `versionKey` and will be assumed to be v0. You'll need to determine in `NewDB` whether you have actually created a new file with `bbolt.Open`, and if so store the current `DBVersion` to prevent going through the entire upgrade sequence.